web/src/components/navbar/NavBarPopupMenu.test.tsx

```diff
@@ -0,0 +1,281 @@
+import {
+  BusinessUser,
+  generateProfileData,
+  generateUser,
+  generateUserData,
+  UserData,
+} from "@businessnjgovnavigator/shared";
+import { fireEvent, render, screen, waitFor } from "@testing-library/react";
+
+const Config = getMergedConfig();
+
+const mockApi = api as jest.Mocked<typeof api>;
+
+jest.mock("next/router", () => ({ useRouter: jest.fn() }));
+jest.mock("@/lib/data-hooks/useUserData", () => ({ useUserData: jest.fn() }));
+jest.mock("@/lib/data-hooks/useRoadmap", () => ({ useRoadmap: jest.fn() }));
+jest.mock("@/lib/api-client/apiClient", () => ({ postSelfReg: jest.fn() }));
+jest.mock("@/lib/auth/signinHelper", () => {
+  const originalSigninHelper = jest.requireActual("@/lib/auth/signinHelper");
+  return {
+    ...originalSigninHelper,
+    onSignOut: jest.fn(),
+  };
+});
+jest.mock("@/lib/auth/sessionHelper", () => ({
+  triggerSignIn: jest.fn(),
+}));
+
+const generateOnboardingUser = (): UserData => {
+  return generateUserData({
+    profileData: generateProfileData({
+      businessName: "",
+      industryId: undefined,
+      legalStructureId: undefined,
+    }),
+  });
+};
+
+const businessName = "businessName";
+
+const generateBusinessNamedUserData = (overrides?: Partial<UserData>): UserData => {
+  return generateUserData({
+    profileData: generateProfileData({
+      businessName: businessName,
+      industryId: "cannabis",
+      legalStructureId: "limited-liability-company",
+    }),
+    ...overrides,
+  });
+};
+
+describe("<NavBarPopupMenu />", () => {
+  beforeEach(() => {
+    jest.resetAllMocks();
+    useMockRouter({});
+  });
+
+  const renderNavBarPopupMenu = (
+    overrides: Partial<NavBarPopupMenuProps>,
+    userData?: UserData,
+    user?: BusinessUser
+  ): void => {
+    render(
+      withAuth(
+        <WithStatefulUserData initialUserData={userData}>
+          <NavBarPopupMenu
+            handleClose={(): void => {}}
+            open={true}
+            hasCloseButton={false}
+            menuConfiguration={"login"}
+            {...overrides}
+          />
+        </WithStatefulUserData>,
+        { user, isAuthenticated: IsAuthenticated.FALSE }
+      )
+    );
+  };
+
+  it("renders close button when hasCloseButton prop is true", () => {
+    useMockUserData({});
+    renderNavBarPopupMenu({ hasCloseButton: true });
+    expect(screen.getByTestId("close-button-nav-menu")).toBeInTheDocument();
+  });
+
+  it("triggers handleClose when close button is pressed", () => {
+    useMockUserData({});
+    renderNavBarPopupMenu({ hasCloseButton: false });
+    expect(screen.queryByTestId("close-button-nav-menu")).not.toBeInTheDocument();
+  });
+
+  describe("menuConfiguration prop's various menu configurations", () => {
+    describe("login configuration", () => {
+      it("shows login button", () => {
+        useMockUserData(generateOnboardingUser());
+        renderNavBarPopupMenu({ menuConfiguration: "login" });
```

Comment:
So we have a weird situation here where these 2 lines are repeated a a few times each and I considered consolidating them but they each or making them one bigger function but render calls can only happen in it or describe blocks and beyond that I'm not really sure we gain much from doing so.

Reply:
Similar situation with some of the checks I'm making in that we could consolidate but it feels to me like they don't actually happen that many times and it wouldn't increase.

Many of these checks happening 3 times which I think is perhaps in my range of acceptable copying? 

4 should be a function, 2 should be copied if right next to one another, 3 is a gray area for me? Curious as to others opinions

Reply:
this seems like an excellent situation to put these 2 (or even 3) lines in a beforeEach for the "login configuration" describe block